lib/rucio/client/downloadclient.py

```diff
@@ -73,6 +73,134 @@ def __init__(self, _client=None, user_agent='rucio_clients', logger=None):
+            did                 - DID string of the archive file (e.g. 'scope:file.name'). Wildcards are not allowed
+            rse                 - rse name (e.g. 'CERN-PROD_DATADISK'). RSE Expressions are not allowed
+            archive             - name of the archive from which the file should be extracted
+            base_dir            - Optional: Base directory where the downloaded files will be stored. (Default: '.')
+            no_subdir           - Optional: If true, files are written directly into base_dir and existing files are overwritten. (Default: False)
+            ignore_checksum     - Optional: If true, the checksum validation is skipped (for pfn downloads the checksum must be given explicitly). (Default: True)
+            transfer_timeout    - Optional: Timeout time for the download protocols. (Default: None)
+        :param num_threads: Suggestion of number of threads to use for the download. It will be lowered if it's too high.
+        :param trace_custom_fields: Custom key value pairs to send with the traces
+
+        :returns: a list of dictionaries with an entry for each file, containing the input options, the did, and the clientState
+                  clientState can be one of the following: ALREADY_DONE, DONE, FILE_NOT_FOUND, FAIL_VALIDATE, FAILED
+
+        :raises InputValidationError: if one of the input items is in the wrong format
+        :raises NoFilesDownloaded: if no files could be downloaded
+        :raises NotAllFilesDownloaded: if not all files could be downloaded
+        :raises RucioException: if something unexpected went wrong during the download
+        """
+        logger = self.logger
+        trace = copy.deepcopy(self.trace_tpl)
+        log_prefix = 'Extracting'
+
+        logger.info('Processing %d item(s) for input' % len(items))
+        input_items = []
+        for item in items:
+            archive = item.get('archive')
+            file_extract = item.get('did')
+            rse_name = item.get('rse')
+            file_extract_scope, file_extract_name = self._split_did_str(file_extract)
+            archive_scope, archive_name = self._split_did_str(archive)
+
+            # if not specified, taking first availbale replica
+            if not rse_name:
+                replicas = self.client.list_replicas([{'scope': archive_scope, 'name': archive_name}],
+                                                     schemes='root',
+                                                     client_location=detect_client_location())
+                for r in replicas:
+                    for rse in r['states'].keys():
+                        if r['states'][rse] == 'AVAILABLE':
+                            rse_attr = self.client.list_rse_attributes(rse)
+                            if rse_attr['istape'] == 'False':
+                                rse_name = rse
+                                break
+                    if rse_name and rse_name is not None:
+                        break
+
```

Comment:
You can use list_replicas with `unavailable=False` to exclude unavailable replicas
and set an rse_expression to `'istape=False' if not rse else '(%s)&istape=False' % rse` to exclude tape endpoints

Reply:
You are not using a site state here are you?
I'm pretty sure `r['states'][rse]` is related to the replica `r`
And `unavailable=False` also:
https://github.com/rucio/rucio/blob/master/lib/rucio/client/replicaclient.py#L101

Reply:
If rse_name is defined, I am not calling list_replicas at all. Thus I left only:
rse_expression='istape=False'